File: rust/tests/test_vectors.rs
extern crate data_encoding;
extern crate serde_json;

use self::data_encoding::HEXLOWER;
pub use self::serde_json::Value as JsonValue;
use std::fs::File;
use std::io::Read;
use std::path::Path;

/// AES block cipher test vectors
// TODO: switch to the tjson crate (based on serde)
#[derive(Debug)]
pub struct AesExample {
    pub key: Vec<u8>,
    pub src: Vec<u8>,
    pub dst: Vec<u8>,
}

impl AesExample {
    /// Load examples from aes.tjson
    pub fn load_all() -> Vec<Self> {
        Self::load_from_file(Path::new("../vectors/aes.tjson"))
    }

    /// Load examples from a file at the given path
    pub fn load_from_file(path: &Path) -> Vec<Self> {
        let mut file = File::open(&path).expect("valid aes.tjson");
        let mut tjson_string = String::new();
        file.read_to_string(&mut tjson_string).expect(
            "aes.tjson read successfully",
        );

        let tjson: serde_json::Value =
            serde_json::from_str(&tjson_string).expect("aes.tjson parses successfully");
        let examples = &tjson["examples:A<O>"].as_array().expect(
            "aes.tjson examples array",
        );

        examples
            .into_iter()
            .map(|ex| {
                Self {
                    key: HEXLOWER
                        .decode(ex["key:d16"].as_str().expect("encoded example").as_bytes())
                        .expect("hex encoded"),
                    src: HEXLOWER
                        .decode(ex["src:d16"].as_str().expect("encoded example").as_bytes())
                        .expect("hex encoded"),
                    dst: HEXLOWER
                        .decode(ex["dst:d16"].as_str().expect("encoded example").as_bytes())
                        .expect("hex encoded"),
                }
            })
            .collect()
    }
}

/// AES-CMAC test vectors
// TODO: switch to the tjson crate (based on serde)
#[derive(Debug)]
pub struct AesCmacExample {
    pub key: Vec<u8>,
    pub message: Vec<u8>,
    pub tag: Vec<u8>,
}

impl AesCmacExample {
    /// Load examples from aes_cmac.tjson
    pub fn load_all() -> Vec<Self> {
        Self::load_from_file(Path::new("../vectors/aes_cmac.tjson"))
    }

    /// Load examples from a file at the given path
    pub fn load_from_file(path: &Path) -> Vec<Self> {
        let mut file = File::open(&path).expect("valid aes_cmac.tjson");
        let mut tjson_string = String::new();
        file.read_to_string(&mut tjson_string).expect(
            "aes_cmac.tjson read successfully",
        );

        let tjson: serde_json::Value =
            serde_json::from_str(&tjson_string).expect("aes_cmac.tjson parses successfully");
        let examples = &tjson["examples:A<O>"].as_array().expect(
            "aes_cmac.tjson examples array",
        );

        examples
            .into_iter()
            .map(|ex| {
                Self {
                    key: HEXLOWER
                        .decode(ex["key:d16"].as_str().expect("encoded example").as_bytes())
                        .expect("hex encoded"),
                    message: HEXLOWER
                        .decode(
                            ex["message:d16"]
                                .as_str()
                                .expect("encoded example")
                                .as_bytes(),
                        )
                        .expect("hex encoded"),
                    tag: HEXLOWER
                        .decode(ex["tag:d16"].as_str().expect("encoded example").as_bytes())
                        .expect("hex encoded"),
                }
            })
            .collect()
    }
}

/// AES-CTR test vectors
// TODO: switch to the tjson crate (based on serde)
#[derive(Debug)]
pub struct AesCtrExample {
    pub key: Vec<u8>,
    pub iv: Vec<u8>,
    pub plaintext: Vec<u8>,
    pub ciphertext: Vec<u8>,
}

impl AesCtrExample {
    /// Load examples from aes_ctr.tjson
    pub fn load_all() -> Vec<Self> {
        Self::load_from_file(Path::new("../vectors/aes_ctr.tjson"))
    }

    /// Load examples from a file at the given path
    pub fn load_from_file(path: &Path) -> Vec<Self> {
        let mut file = File::open(&path).expect("valid aes_ctr.tjson");
        let mut tjson_string = String::new();
        file.read_to_string(&mut tjson_string).expect(
            "aes_ctr.tjson read successfully",
        );

        let tjson: serde_json::Value =
            serde_json::from_str(&tjson_string).expect("aes_ctr.tjson parses successfully");
        let examples = &tjson["examples:A<O>"].as_array().expect(
            "aes_ctr.tjson examples array",
        );

        examples
            .into_iter()
            .map(|ex| {
                Self {
                    key: HEXLOWER
                        .decode(ex["key:d16"].as_str().expect("encoded example").as_bytes())
                        .expect("hex encoded"),
                    iv: HEXLOWER
                        .decode(ex["iv:d16"].as_str().expect("encoded example").as_bytes())
                        .expect("hex encoded"),
                    plaintext: HEXLOWER
                        .decode(
                            ex["plaintext:d16"]
                                .as_str()
                                .expect("encoded example")
                                .as_bytes(),
                        )
                        .expect("hex encoded"),
                    ciphertext: HEXLOWER
                        .decode(
                            ex["ciphertext:d16"]
                                .as_str()
                                .expect("encoded example")
                                .as_bytes(),
                        )
                        .expect("hex encoded"),
                }
            })
            .collect()
    }
}

/// AES-PMAC test vectors
// TODO: switch to the tjson crate (based on serde)
#[derive(Debug)]
pub struct AesPmacExample {
    pub key: Vec<u8>,
    pub message: Vec<u8>,
    pub tag: Vec<u8>,
}

impl AesPmacExample {
    /// Load examples from aes_pmac.tjson
    pub fn load_all() -> Vec<Self> {
        Self::load_from_file(Path::new("../vectors/aes_pmac.tjson"))
    }

    /// Load examples from a file at the given path
    pub fn load_from_file(path: &Path) -> Vec<Self> {
        let mut file = File::open(&path).expect("valid aes_pmac.tjson");
        let mut tjson_string = String::new();
        file.read_to_string(&mut tjson_string).expect(
            "aes_pmac.tjson read successfully",
        );

        let tjson: serde_json::Value =
            serde_json::from_str(&tjson_string).expect("aes_pmac.tjson parses successfully");
        let examples = &tjson["examples:A<O>"].as_array().expect(
            "aes_pmac.tjson examples array",
        );

        examples
            .into_iter()
            .map(|ex| {
                Self {
                    key: HEXLOWER
                        .decode(ex["key:d16"].as_str().expect("encoded example").as_bytes())
                        .expect("hex encoded"),
                    message: HEXLOWER
                        .decode(
                            ex["message:d16"]
                                .as_str()
                                .expect("encoded example")
                                .as_bytes(),
                        )
                        .expect("hex encoded"),
                    tag: HEXLOWER
                        .decode(ex["tag:d16"].as_str().expect("encoded example").as_bytes())
                        .expect("hex encoded"),
                }
            })
            .collect()
    }
}

/// AES-SIV test vectors
// TODO: switch to the tjson crate (based on serde)
#[derive(Debug)]
pub struct AesSivExample {
    pub key: Vec<u8>,
    pub ad: Vec<Vec<u8>>,
    pub plaintext: Vec<u8>,
    pub ciphertext: Vec<u8>,
}

impl AesSivExample {
    /// Load examples from aes_siv.tjson
    pub fn load_all() -> Vec<Self> {
        Self::load_from_file(Path::new("../vectors/aes_siv.tjson"))
    }

    /// Load examples from a file at the given path
    pub fn load_from_file(path: &Path) -> Vec<Self> {
        let mut file = File::open(&path).expect("valid aes_siv.tjson");
        let mut tjson_string = String::new();
        file.read_to_string(&mut tjson_string).expect(
            "aes_siv.tjson read successfully",
        );

        let tjson: serde_json::Value =
            serde_json::from_str(&tjson_string).expect("aes_siv.tjson parses successfully");
        let examples = &tjson["examples:A<O>"].as_array().expect(
            "aes_siv.tjson examples array",
        );

        examples
            .into_iter()
            .map(|ex| {
                Self {
                    key: HEXLOWER
                        .decode(ex["key:d16"].as_str().expect("encoded example").as_bytes())
                        .expect("hex encoded"),
                    ad: ex["ad:A<d16>"]
                        .as_array()
                        .expect("encoded example")
                        .iter()
                        .map(|ex| {
                            HEXLOWER
                                .decode(ex.as_str().expect("encoded example").as_bytes())
                                .expect("hex encoded")
                        })
                        .collect(),
                    plaintext: HEXLOWER
                        .decode(
                            ex["plaintext:d16"]
                                .as_str()
                                .expect("encoded example")
                                .as_bytes(),
                        )
                        .expect("hex encoded"),
                    ciphertext: HEXLOWER
                        .decode(
                            ex["ciphertext:d16"]
                                .as_str()
                                .expect("encoded example")
                                .as_bytes(),
                        )
                        .expect("hex encoded"),
                }
            })
            .collect()
    }
}

/// AES-SIV test vectors
// TODO: switch to the tjson crate (based on serde)
#[derive(Debug)]
pub struct AesPmacSivExample {
    pub key: Vec<u8>,
    pub ad: Vec<Vec<u8>>,
    pub plaintext: Vec<u8>,
    pub ciphertext: Vec<u8>,
}

impl AesPmacSivExample {
    /// Load examples from aes_pmac_siv.tjson
    pub fn load_all() -> Vec<Self> {
        Self::load_from_file(Path::new("../vectors/aes_pmac_siv.tjson"))
    }

    /// Load examples from a file at the given path
    pub fn load_from_file(path: &Path) -> Vec<Self> {
        let mut file = File::open(&path).expect("valid aes_pmac_siv.tjson");
        let mut tjson_string = String::new();
        file.read_to_string(&mut tjson_string).expect(
            "aes_pmac_siv.tjson read successfully",
        );

        let tjson: serde_json::Value =
            serde_json::from_str(&tjson_string).expect("aes_pmac_siv.tjson parses successfully");
        let examples = &tjson["examples:A<O>"].as_array().expect(
            "aes_pmac_siv.tjson examples array",
        );

        examples
            .into_iter()
            .map(|ex| {
                Self {
                    key: HEXLOWER
                        .decode(ex["key:d16"].as_str().expect("encoded example").as_bytes())
                        .expect("hex encoded"),
                    ad: ex["ad:A<d16>"]
                        .as_array()
                        .expect("encoded example")
                        .iter()
                        .map(|ex| {
                            HEXLOWER
                                .decode(ex.as_str().expect("encoded example").as_bytes())
                                .expect("hex encoded")
                        })
                        .collect(),
                    plaintext: HEXLOWER
                        .decode(
                            ex["plaintext:d16"]
                                .as_str()
                                .expect("encoded example")
                                .as_bytes(),
                        )
                        .expect("hex encoded"),
                    ciphertext: HEXLOWER
                        .decode(
                            ex["ciphertext:d16"]
                                .as_str()
                                .expect("encoded example")
                                .as_bytes(),
                        )
                        .expect("hex encoded"),
                }
            })
            .collect()
    }
}

/// dbl() test vectors
// TODO: switch to the tjson crate (based on serde)
#[derive(Debug)]
pub struct DblExample {
    pub input: Vec<u8>,
    pub output: Vec<u8>,
}

impl DblExample {
    /// Load examples from dbl.tjson
    pub fn load_all() -> Vec<Self> {
        Self::load_from_file(Path::new("../vectors/dbl.tjson"))
    }

    /// Load examples from a file at the given path
    pub fn load_from_file(path: &Path) -> Vec<Self> {
        let mut file = File::open(&path).expect("valid dbl.tjson");
        let mut tjson_string = String::new();
        file.read_to_string(&mut tjson_string).expect(
            "dbl.tjson read successfully",
        );

        let tjson: serde_json::Value =
            serde_json::from_str(&tjson_string).expect("dbl.tjson parses successfully");
        let examples = &tjson["examples:A<O>"].as_array().expect(
            "dbl.tjson examples array",
        );

        examples
            .into_iter()
            .map(|ex| {
                Self {
                    input: HEXLOWER
                        .decode(
                            ex["input:d16"]
                                .as_str()
                                .expect("encoded example")
                                .as_bytes(),
                        )
                        .expect("hex encoded"),
                    output: HEXLOWER
                        .decode(
                            ex["output:d16"]
                                .as_str()
                                .expect("encoded example")
                                .as_bytes(),
                        )
                        .expect("hex encoded"),
                }
            })
            .collect()
    }
}

File: js/test/support/test_vectors.ts
import * as fs from "async-file";
import TJSON from "tjson-js";

/** AES-SIV test vectors */
export class AesSivExample {
  static readonly DEFAULT_EXAMPLES_PATH = "../vectors/aes_siv.tjson";

  public readonly name: string;
  public readonly key: Uint8Array;
  public readonly ad: Uint8Array[];
  public readonly plaintext: Uint8Array;
  public readonly ciphertext: Uint8Array;

  static async loadAll(): Promise<AesSivExample[]> {
    return AesSivExample.loadFromFile(AesSivExample.DEFAULT_EXAMPLES_PATH);
  }

  static async loadFromFile(filename: string): Promise<AesSivExample[]> {
    let tjson = TJSON.parse(await fs.readFile(filename, "utf8"));
    return tjson["examples"].map((ex: any) => {
      let obj = Object.create(AesSivExample.prototype);
      return Object.assign(obj, ex);
    });
  }
}

/** AES-PMAC-SIV test vectors */
export class AesPmacSivExample {
  static readonly DEFAULT_EXAMPLES_PATH = "../vectors/aes_pmac_siv.tjson";

  public readonly name: string;
  public readonly key: Uint8Array;
  public readonly ad: Uint8Array[];
  public readonly plaintext: Uint8Array;
  public readonly ciphertext: Uint8Array;

  static async loadAll(): Promise<AesPmacSivExample[]> {
    return AesPmacSivExample.loadFromFile(AesPmacSivExample.DEFAULT_EXAMPLES_PATH);
  }

  static async loadFromFile(filename: string): Promise<AesPmacSivExample[]> {
    let tjson = TJSON.parse(await fs.readFile(filename, "utf8"));
    return tjson["examples"].map((ex: any) => {
      let obj = Object.create(AesPmacSivExample.prototype);
      return Object.assign(obj, ex);
    });
  }
}

/** AEAD (AES-SIV/AES-PMAC-SIV) test vectors */
export class AEADExample {
  static readonly DEFAULT_EXAMPLES_PATH = "../vectors/aes_siv_aead.tjson";

  public readonly name: string;
  public readonly alg: string;
  public readonly key: Uint8Array;
  public readonly ad: Uint8Array;
  public readonly nonce: Uint8Array;
  public readonly plaintext: Uint8Array;
  public readonly ciphertext: Uint8Array;

  static async loadAll(): Promise<AEADExample[]> {
    return AEADExample.loadFromFile(AEADExample.DEFAULT_EXAMPLES_PATH);
  }

  static async loadFromFile(filename: string): Promise<AEADExample[]> {
    let tjson = TJSON.parse(await fs.readFile(filename, "utf8"));
    return tjson["examples"].map((ex: any) => {
      let obj = Object.create(AEADExample.prototype);
      return Object.assign(obj, ex);
    });
  }
}

export class STREAMBlock {
  public readonly ad: Uint8Array;
  public readonly plaintext: Uint8Array;
  public readonly ciphertext: Uint8Array;
}

/** STREAM (AES-SIV/AES-PMAC-SIV) test vectors */
export class STREAMExample {
  static readonly DEFAULT_EXAMPLES_PATH = "../vectors/aes_siv_stream.tjson";

  public readonly name: string;
  public readonly alg: string;
  public readonly key: Uint8Array;
  public readonly nonce: Uint8Array;
  public readonly blocks: STREAMBlock[];

  static async loadAll(): Promise<STREAMExample[]> {
    return STREAMExample.loadFromFile(STREAMExample.DEFAULT_EXAMPLES_PATH);
  }

  static async loadFromFile(filename: string): Promise<STREAMExample[]> {
    let tjson = TJSON.parse(await fs.readFile(filename, "utf8"));
    return tjson["examples"].map((ex: any) => {
      let obj = Object.create(STREAMExample.prototype);
      return Object.assign(obj, ex);
    });
  }
}

/** AES (raw block function) test vectors */
export class AesExample {
  static readonly DEFAULT_EXAMPLES_PATH = "../vectors/aes.tjson";

  public readonly key: Uint8Array;
  public readonly src: Uint8Array;
  public readonly dst: Uint8Array;

  static async loadAll(): Promise<AesExample[]> {
    return AesExample.loadFromFile(AesExample.DEFAULT_EXAMPLES_PATH);
  }

  static async loadFromFile(filename: string): Promise<AesExample[]> {
    let tjson = TJSON.parse(await fs.readFile(filename, "utf8"));
    return tjson["examples"].map((ex: any) => {
      let obj = Object.create(AesExample.prototype);
      return Object.assign(obj, ex);
    });
  }
}

/** AES-CTR test vectors */
export class AesCtrExample {
  static readonly DEFAULT_EXAMPLES_PATH = "../vectors/aes_ctr.tjson";

  public readonly key: Uint8Array;
  public readonly iv: Uint8Array;
  public readonly plaintext: Uint8Array;
  public readonly ciphertext: Uint8Array;

  static async loadAll(): Promise<AesCtrExample[]> {
    return AesCtrExample.loadFromFile(AesCtrExample.DEFAULT_EXAMPLES_PATH);
  }

  static async loadFromFile(filename: string): Promise<AesCtrExample[]> {
    let tjson = TJSON.parse(await fs.readFile(filename, "utf8"));
    return tjson["examples"].map((ex: any) => {
      let obj = Object.create(AesCtrExample.prototype);
      return Object.assign(obj, ex);
    });
  }
}

/** AES-CMAC test vectors */
export class AesCmacExample {
  static readonly DEFAULT_EXAMPLES_PATH = "../vectors/aes_cmac.tjson";

  public readonly key: Uint8Array;
  public readonly message: Uint8Array;
  public readonly tag: Uint8Array;

  static async loadAll(): Promise<AesCmacExample[]> {
    return AesCmacExample.loadFromFile(AesCmacExample.DEFAULT_EXAMPLES_PATH);
  }

  static async loadFromFile(filename: string): Promise<AesCmacExample[]> {
    let tjson = TJSON.parse(await fs.readFile(filename, "utf8"));
    return tjson["examples"].map((ex: any) => {
      let obj = Object.create(AesCmacExample.prototype);
      return Object.assign(obj, ex);
    });
  }
}

/** AES-PMAC test vectors */
export class AesPmacExample {
  static readonly DEFAULT_EXAMPLES_PATH = "../vectors/aes_pmac.tjson";

  public readonly key: Uint8Array;
  public readonly message: Uint8Array;
  public readonly tag: Uint8Array;

  static async loadAll(): Promise<AesPmacExample[]> {
    return AesPmacExample.loadFromFile(AesPmacExample.DEFAULT_EXAMPLES_PATH);
  }

  static async loadFromFile(filename: string): Promise<AesPmacExample[]> {
    let tjson = TJSON.parse(await fs.readFile(filename, "utf8"));
    return tjson["examples"].map((ex: any) => {
      let obj = Object.create(AesPmacExample.prototype);
      return Object.assign(obj, ex);
    });
  }
}

/** dbl() test vectors */
export class DblExample {
  static readonly DEFAULT_EXAMPLES_PATH = "../vectors/dbl.tjson";

  public readonly input: Uint8Array;
  public readonly output: Uint8Array;

  static async loadAll(): Promise<DblExample[]> {
    return DblExample.loadFromFile(DblExample.DEFAULT_EXAMPLES_PATH);
  }

  static async loadFromFile(filename: string): Promise<DblExample[]> {
    let tjson = TJSON.parse(await fs.readFile(filename, "utf8"));
    return tjson["examples"].map((ex: any) => {
      let obj = Object.create(DblExample.prototype);
      return Object.assign(obj, ex);
    });
  }
}
